backend/wmg/pipeline/utils.py

```diff
@@ -72,22 +85,38 @@ def anndata_filter_cells_by_gene_counts_inplace(adata: AnnData, min_genes: int)
         # That is, ROW 3 is not stored in the `adata.X.indices` array:
         # column_oriented_sparse_matrix = np.array([[1, 0, 0, 2], [0, 4, 1, 0], [0, 0, 0, 0], [0, 0, 5, 0]])
         logger.info(f"Filtering cells for anndata.X csc_matrix of shape: {adata.X.shape}")
-        number_per_cell = np.zeros(adata.shape[0], dtype="int")
-        number_per_cell[cell_indices] = counts
+        gene_counts_per_cell = np.zeros(adata.shape[0], dtype="int")
+        gene_counts_per_cell[cell_indices] = counts
     elif isinstance(adata.X, np.ndarray):
         logger.info(f"Filtering cells for anndata.X ndarray of shape: {adata.X.shape}")
-        number_per_cell = (adata.X > 0).sum(axis=1).flatten()
+        gene_counts_per_cell = (adata.X > 0).sum(axis=1).flatten()
     else:
         raise UnsupportedMatrixTypeError(f"Unsupported AnnData.X matrix type: {type(adata.X)}")
 
-    logger.info(f"Shape of number_per_cell array holding counts of non-zero column values: {number_per_cell.shape}")
-    cell_subset = number_per_cell >= min_genes
+    logger.info(
+        f"Shape of gene_counts_per_cell array holding counts of non-zero column values: {gene_counts_per_cell.shape}"
+    )
+
+    # boolean index encoding the cells (rows) with number of expressed genes less than the threshold
+    # and that should be marked to be filtered out
+    lowly_covered_cells_to_filter_out = gene_counts_per_cell < min_genes
+
+    adata.obs["filter_cells"] = lowly_covered_cells_to_filter_out
+
+    # Do inplace zeroing of rows that should be filtered out
+    # This is done to make the anndata object sparser and memory efficient
+    if lowly_covered_cells_to_filter_out.sum() > 0:
+        logger.info("Performing in place zeroing of AnnData.X rows to be filtered out")
+        rows_to_zero = np.where(lowly_covered_cells_to_filter_out)[0]
+        if isinstance(adata.X, csr_matrix):
+            for row in rows_to_zero:
+                adata.X.data[adata.X.indptr[row] : adata.X.indptr[row + 1]] = 0
+            adata.X.eliminate_zeros()
+        elif isinstance(adata.X, csc_matrix):
+            adata.X.data[np.in1d(adata.X.indices, rows_to_zero)] = 0
+            adata.X.eliminate_zeros()
+        else:
```

Comment:
Just for safety, we should probably have `elif isinstance(adata.X, np.ndarray)` and then `else` raise invalid matrix type exception

Reply:
will fix